homeassistant/components/axis.py

```diff
@@ -0,0 +1,347 @@
+    vol.Optional(CONF_SCAN_INTERVAL, default=0): cv.positive_int,
+    vol.Optional(ATTR_LOCATION, default=None): cv.string,
+}, extra=vol.ALLOW_EXTRA)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        cv.slug: DEVICE_SCHEMA,
+    }),
+}, required=False, extra=vol.ALLOW_EXTRA)
+
+
+def request_configuration(hass, name, host, serialnumber):
+    """Request configuration steps from the user."""
+    configurator = get_component('configurator')
+
+    def configuration_callback(callback_data):
+        """Called when config is submitted."""
+        if CONF_INCLUDE not in callback_data:
+            configurator.notify_errors(instance,
+                                       'Functionality mandatory.')
+            return False
+        callback_data[CONF_INCLUDE] = callback_data[CONF_INCLUDE].split()
+        callback_data[CONF_HOST] = host
+        if CONF_NAME not in callback_data:
+            callback_data[CONF_NAME] = name
+        try:
+            config = DEVICE_SCHEMA(callback_data)
+        except:  # pylint: disable=W0702
+            configurator.notify_errors(instance,
+                                       'Bad input, please check spelling.')
+            return False
+
+        config['hass'] = hass
+
+        if setup_device(config):
+            config_file = _read_config(hass)
+            config_file[serialnumber] = dict(config)
+            del config_file[serialnumber]['hass']
+            _write_config(hass, config_file)
+            configurator.request_done(instance)
+        else:
+            configurator.notify_errors(instance,
+                                       'Failed to register, please try again.')
+            return False
+
+    title = '{} ({})'.format(name, host)
+    instance = configurator.request_config(
+        hass, title, configuration_callback,
+        description='Functionality: ' + str(AXIS_INCLUDE),
+        entity_picture="/static/images/logo_axis.png",
+        link_name='device live view',
+        link_url='http://' + host,
+        submit_caption="Confirm",
+        fields=[
+            {'id': CONF_NAME,
+             'name': 'Device name',
+             'type': 'text'},
+            {'id': CONF_USERNAME,
+             'name': 'User name',
+             'type': 'text'},
+            {'id': CONF_PASSWORD,
+             'name': 'Password',
+             'type': 'password'},
+            {'id': CONF_INCLUDE,
+             'name': 'Device functionality (space separated list)',
+             'type': 'text'},
+            {'id': ATTR_LOCATION,
+             'name': 'Physical location of device (optional)',
+             'type': 'text'},
+            {'id': CONF_SCAN_INTERVAL,
+             'name': 'Sensor update interval (optional)',
+             'type': 'number'},
+        ]
+    )
+
+
+def setup(hass, base_config):
+    """Common setup for Axis devices."""
+    configurator = get_component('configurator')
+    configurator._get_instance(hass)  # pylint: disable=W0212
```

Comment:
- Regarding keeping track of request_id, I'm doing that, but will rename instance to request_id for clarification.
- What I've seen here now is that every discovery entering `request_configuration` doing their `request_config` will in the `_get_instance` function always get `hass.data.get(_KEY_INSTANCE) == None`
- This means that for every discovery in axis platform `hass.data[_KEY_INSTANCE] = Configurator(hass)` will be run.
- This means that request_id will always contain a unique id of 1 instead of 2, 3, 4, etc.
- When pushing the `submit button` in the GUI, will always try to continue using the first request_id to be initiated. Which means trying to configure devices not discovered first, will always fail.
**Faulty call**
- 17-05-05 14:47:48 INFO (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=configurator, service_call_id=1964134384-1, service=configure, service_data=configure_id=1900023920-1, fields=username=root, password=pass, include=camera pir, name=1234>
17-05-05 14:47:48 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_executed[L]: service_call_id=1964134384-1>
**Working call**
- 17-05-05 14:48:52 INFO (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=configurator, service_call_id=1964134384-2, service=configure, service_data=configure_id=1900024272-1, fields=username=root, password=pass, include=camera pir, name=1234>
Enter configuration callback {'username': 'root', 'password': 'pass', 'include': 'camera pir', 'name': '1234'}
17-05-05 14:48:52 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_executed[L]: service_call_id=1964134384-2>
Only difference between these print outs are the configure_id, but one device gets configured correctly and the other one doesn't even enter the configuration callback

So I still think that I'm doing it a proper way :)